src/common/cockpitwebcertificate.c

```diff
@@ -112,6 +112,30 @@ cockpit_certificate_locate (char **error)
   return NULL;
 }
 
+/**
+ * cockpit_certificate_key_path:
+ *
+ * Return key file path for given certfile, i. e. replace ".crt" or ".cert"
+ * suffix with ".key". Return %NULL for invalid names.
+ */
+char *
+cockpit_certificate_key_path (const char *certfile)
+{
+  int len = strlen (certfile);
+  char *keypath = NULL;
+
+  /* .cert suffix case: chop off suffix, append ".key" */
+  if (len > 5 && strcmp (certfile + len - 5, ".cert") == 0)
+    asprintfx (&keypath, "%.*s.key", len - 5, certfile);
+  /* *.crt suffix case */
+  else if (len > 4 && strcmp (certfile + len - 4, ".crt") == 0)
+    asprintfx (&keypath, "%.*s.key", len - 4, certfile);
+  else
+    warnx ("invalid certificate file name: %s", certfile);
```

Comment:
This should be an assert, shouldn't it?  Is there any way that this might actually happen?

Reply:
Good point -- we currently call it from cockpit_certificate_load() and connection_crypto_init(), both of which currently accept arbitary cert paths. connection_crypto_init() is fine, it gets its input from cockpit_certificate_locate() which only spits out .crt/.cert. Same in src/ws/main.c.

I was a bit worried about remotectl when calling it with several PEM files. But cockpit_certificate_combine() fixes the target path to *.cert", and only that gets loaded, so it should be fine.

Reply:
Changed to `errx()`, to get an useful message.